### PR TITLE
feat(cli): migrate to zero

### PR DESCRIPTION
**Pull request by bodinsamuel**
## Changes

Fixes https://linear.app/nango/issue/NAN-3245/migration-command

- `nango migrate-to-zero-yaml` 
Add massive script to migrate full codebase to new syntax. On best case it migrates everything without a single error, in most cases there are a few types error at the end (e.g: import leftover, unstrict typings that becomes invalid).
Tested on a few customers with great success but it's probably missing a few edge cases.

- Change export to import in index.ts
I realized there was some TS conflict because they all export the same things, it doesn't really matter since I'm not compiling this file but merely using it has declaration.

- Try to use cli-test setup to fix windows test
Not sure if it's that but worth the shot, otherwise I'll create another PR for the fix

## Tests

> [!CAUTION]
> This operation is destructive, git commit or save your folder before testing

- `node packages/cli/dist/index.js migrate-to-zero-yaml`
- ???
- Profit

---

**Add CLI Zero-YAML Migration Command: Full Codebase AST Refactor, Zod Model Generation, and TypeScript-Only Integrations**

This PR introduces a major CLI migration feature: `nango migrate-to-zero-yaml`. It provides an automated, CLI-driven transformation of legacy integration codebases using the nango.yaml+script convention to the new Zero-YAML, TypeScript/Zod-only architecture. The migration script runs extensive AST-based refactoring-rebuilding all sync/action/on-event scripts, replacing YAML with models.ts (generated Zod schemas), rewriting the import/export structure for ESM/CJS compatibility, and updating package.json/tsconfig to fit new standard practices. It enforces improved runner SDK typings, especially around batch operations and model handling. Snapshots and unit tests are included for transformation validation and edge case coverage. The migration is intentionally destructive and must be run with backups in place.

**Key Changes:**
• Introduces the `nango migrate-to-zero-yaml` ``CLI`` command for automating migration from nango.yaml+scripts to `TypeScript`/Zod-based Zero-``YAML`` structure.
• Implements large-scale ``AST``/code transformation using jscodeshift: rewrites all script types (syncs/actions/events/helpers), regenerates imports/exports, and introduces stricter `ZodModel` enforcement.
• Generates models.ts with topologically sorted Zod models/types from legacy ``YAML`` definitions.
• Deletes nango.yaml, rewrites index.ts to import form, and updates package.json and tsconfig (e.g., includes jscodeshift, tighter exclusion of build artifacts, ensures zod and nango `devDependencies`).
• Refactors runner ``SDK`` type signatures to require all models to define required `id` fields and corrects `batchDelete`/`batchUpdate` to enforce partial record semantics (all fields optional except `id`).
• Adds extensive migration/unit/snapshot tests and helper logic to handle symlinks and edge cases.
• Upgrades handling of helper files: automatically migrates their imports and `NangoSync`/`NangoAction` usage.
• Handles npm install step post-migration to update dependencies.

**Affected Areas:**
• ``CLI`` core/command registry
• packages/cli/lib/migrations/`toZeroYaml`.ts
• Integration project structure (models.ts, index.ts, package.json, tsconfig.json)
• packages/runner-sdk/lib/scripts.ts
• packages/runner-sdk/lib/sync.ts
• All sync/action/on-event/helper scripts
• Unit/snapshot tests for migration
• Zero-``YAML`` compile chain, type definitions

**Potential Impact:**

**Functionality**: Replaces legacy YAML-based integrations with fully TypeScript/Zod-based system. Migrates or overwrites all integration scripts, deletes YAML config, and could introduce type errors requiring manual follow-up.

**Performance**: Migration itself is resource-intensive on large codebases and triggers a full npm install; runtime performance of migrated integrations unaffected.

**Security**: No new security issues, but as migration is destructive, data loss is possible if users don't back up first. Explicitly skips symlinks for safety.

**Scalability**: Designed for robust operation on large/multi-integration repos with dependency-resolution in model generation; edge cases possible for non-standard legacy patterns, requiring occasional manual fixes.

**Review Focus:**
• Correctness and safety of migration ``AST``/code transformations-do converted scripts remain functionally equivalent?
• Does new `batchDelete`/`batchUpdate` type enforcement break or fix intended patterns?
• Import/export and module structure (especially default vs named/export forms in index.ts and scripts).
• Exhaustiveness of test cases and coverage for both typical and edge migration paths.
• Type safety improvements for all Nango runner ``SDK`` usages-are invalid model usages/omissions surfaced?
• Are destructive migration warnings clear and effective for users?

<details>
<summary><strong>Testing Needed</strong></summary>

• Run migration on real-world integration projects (with backup) and verify output `TypeScript` files and models.
• Validate all migrated scripts compile, especially `batchDelete`/`batchUpdate` behavior and presence of `id` in models.
• Check generated models.ts and index.ts imports/exports for correctness.
• Run existing and added unit/snapshot tests in different ``OS`` environments (Linux/Mac/Windows) to ensure portability.
• Test npm install post-migration and confirm updated dependency tree.

</details>

<details>
<summary><strong>Code Quality Assessment</strong></summary>

**packages/runner-sdk/lib/sync.ts**: Stricter model signature, type-safe batch* methods, now enforces id requirement.

**packages/cli/lib/index.ts**: Adds migration command, updates CLI command tree cleanly.

**packages/cli/lib/zeroYaml/definitions.ts**: Adopts new ZodModel constraints, code is dense but logical.

**packages/cli/lib/migrations/toZeroYaml.ts**: Extensive, highly imperative AST manipulation; complex but well-commented for migration scripts, though challenging for full manual audit.

**packages/runner-sdk/lib/scripts.ts**: Improved ZodModel typing, batch method signatures now require 'id' and partial semantics for updates/deletes.

</details>

<details>
<summary><strong>Best Practices</strong></summary>

**Cli**:
• Clear command descriptions, new help text.
• Cleaner error messages, spinner/status feedback.

**Migration**:
• Destructive operations are gated by warnings, backup prompts, and explicit documentation.
• ``AST``-based code transformation for highest fidelity migration.
• Topological sort for Zod model output ensures declaration order matches dependency graph.
• Symlink skipping for file system safety.

**Testing**:
• Unit and snapshot-based testing added for all major transformation paths.
• Cross-platform test recommendations.
• Test mocks for `batchSave` and edge cases.

**Type Safety**:
• Strict `ZodModel` typing enforced for all models used in runner ``SDK``.
• Batch method types now require id for deletes/updates.
• Unifies typing of models, actions, syncs, events across codebase.

</details>

<details>
<summary><strong>Possible Issues</strong></summary>

• Integrations using highly untyped, nonstandard, or symbolic linked scripts may experience partial or failed migration; manual follow-up may be needed.
• Legacy scripts missing required 'id' fields on models will now fail type checks post-migration.
• Migrated batchDelete/batchUpdate signatures could break code that previously deleted by id only if other fields are referenced elsewhere.
• npm install step could cause dependency conflicts if local or custom modifications to package.json/devDependencies exist.
• Some TypeScript errors may remain for unsupportable legacy/edge-case API uses; these are intentionally surfaced for developer correction.
• Destructive migration means failed/interrupted run can leave project in a partially migrated state if not backed up first.

</details>

---
*This summary was automatically generated by @propel-code-bot*

### CODE REVIEW

File: packages/cli/lib/migrations/toZeroYaml.ts [hunk 1]
```diff
@@ -0,0 +1,1146 @@
+import { spawn } from 'node:child_process';
```
Comment (bodinsamuel) [inline packages/cli/lib/migrations/toZeroYaml.ts:1]: I don't except deep review here, it's the fruit of 2weeks of bug chasing and edge cases found in customers scripts. Half of the code is AI slop that I refined manually.
It's massive, tried to breakdown and add comments but it's a bit hard to digest

**Review by TBonnin — COMMENTED** [review]
I haven't reviewed the migration script but I tested it locally. I ran into 2 main breaking changes. There are not due to the migration itself but to the compiler being more strict:
- `nango.batchDelete` used to accept a object with only `id`. Now the compiler expect a full object with all attributes. That might be an issue in scenarios where the full object to delete is not known. Should the type for batchDelete have all fields other than `id` as optional? 
- in a sync derived from [the hubspot prebuilt sync](https://github.com/NangoHQ/integration-templates/blob/main/integrations/hubspot/syncs/contacts.ts) with createdAt field set to type string. Passing a Date used to be allowed. Now the compiler emits an error (Date <> string)

Let me know if you want me to share the code

**Comment by bodinsamuel** [comment]
> nango.batchDelete used to accept a object with only id. Now the compiler expect a full object with all attributes. That might be an issue in scenarios where the full object to delete is not known. Should the type for batchDelete have all fields other than id as optional?

Indeed, I'll change that thanks

> in a sync derived from [the hubspot prebuilt sync](https://github.com/NangoHQ/integration-templates/blob/main/integrations/hubspot/syncs/contacts.ts) with createdAt field set to type string. Passing a Date used to be allowed. Now the compiler emits an error (Date <> string)

I think this one is acceptable, but if you can send me the code just to be sure 👍🏻

**Comment by TBonnin** [comment]
> I think this one is acceptable, but if you can send me the code just to be sure 👍🏻

I agree it is acceptable and quite easy to fix. Here is the code if you want to check: 
```
export async function onWebhookPayloadReceived(
    nango: NangoSync,
    payload: any,
): Promise<void> {
    await nango.log('Webhook payload received', payload);

    const contact = {
        id: SOME_RANDOM_ID,
        created_at: new Date(),
        updated_at: new Date(),
        first_name: 'first',
        last_name: 'last',
        email: 'first.last@email.com',
        active: true,
    };
    await nango.batchSave([contact], 'HubspotContact');
}

```

with the model in nango.yaml.
```
    HubspotContact:
        id: string
        created_at: string
        updated_at: string
        first_name: string
        last_name: string
        email: string

```

**Comment by kaposke** [comment]
Just FYI I'm reviewing and testing as well

**Review by kaposke — COMMENTED** [review]
Might not be with migrations but with zero-yaml itself:
`z.date()` in `models.ts` becomes `date` (invalid lowercase) in `.nango/schema.ts`, which then fails creation of the json schema.
image
image

**Comment by bodinsamuel** [comment]
Thanks @TBonnin @kaposke 
I have fixed the date and the model definitions for batchDelete. Realized I wasn't enforcing an id for the models, so also changed that.
For the model that needs a string but get a date I don't think it's an easy fix, ultimately it's wrong because we will send back a string. I wish JSON supported Date object 😢

**Review by TBonnin — APPROVED** [review]
